onnxruntime/core/providers/cuda/math/einsum.cc

```diff
@@ -0,0 +1,70 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "einsum.h"
+
+namespace onnxruntime {
+
+namespace cuda {
+
+ONNX_OPERATOR_KERNEL_EX(
+    Einsum,
+    kOnnxDomain,
+    12,
+    kCudaExecutionProvider,
+    KernelDefBuilder().TypeConstraint("T",
+                                      std::vector<MLDataType>{
+                                          DataTypeImpl::GetTensorType<float>(),
+                                          DataTypeImpl::GetTensorType<double>()}),
+    Einsum);
+
+Status Einsum::Compute(OpKernelContext* context) const {
+  return onnxruntime::Einsum::Compute(context);
+}
+
+Status Einsum::DeviceCompute(OpKernelContext* context, const std::vector<const Tensor*>& inputs,
+                             AllocatorPtr allocator, concurrency::ThreadPool* tp) const {
+  cublasHandle_t cublas_handle = cuda_ep_->PerThreadCublasHandle();
+
+  // EinsumComputePreprocessor section -
+  auto einsum_compute_preprocessor = EinsumComputePreprocessor(*einsum_equation_preprocessor_, inputs, allocator,
+                                                               cublas_handle);
+  // Set device specific methods (CPU methods) to be used during pre-processing
+  einsum_compute_preprocessor.SetDeviceHelpers(EinsumOp::DeviceHelpers::CudaDeviceHelpers::Diagonal,
```

Comment:
'Set device specific methods (CUDA methods)'?

Reply:
Possibly you don't need the comment here or below. It's fairly self explanatory from the function names (at least to me) what's going on. 

---
In reply to: [433082947](https://github.com/microsoft/onnxruntime/pull/4095#discussion_r433082947) [](ancestors = 433082947)

Reply:
Ok removed the comments from the CPU and CUDA implementations

Reply:
Should it say 'CUDA methods' instead of 'CPU methods'? 

---
In reply to: [441141730](https://github.com/microsoft/onnxruntime/pull/4095#discussion_r441141730) [](ancestors = 441141730)

Reply:
Thought I removed the comments as it was self-explanatory. Removed them now.